jobs/audit_jobs.py

```diff
@@ -0,0 +1,118 @@
+            self.pipeline
+            | 'Get all models' >> self.job_options.model_getter()
+            | 'Partition by model.deleted' >> (
+                beam.Partition(lambda model, _: int(model.deleted), 2))
+        )
+
+        models_of_kind_by_index = (
+            existing_models
+            # NOTE: Partition returns a statically-sized list of PCollections.
+            # Creating partitions is wasteful when there are less items than
+            # there are partitions, like in our unit tests. In exchange, in
+            # production the job will be able to take advantage of the high
+            # parallelizability of PCollections, which are designed for enormous
+            # datasets and parallel processing.
+            #
+            # Alternatively, we could have used GroupBy. However, that returns
+            # an _iterable_ of items rather than a PCollection, and so it is
+            # vulnerable to out-of-memory errors.
+            #
+            # Since this job is concerned with running audits on EVERY MODEL IN
+            # STORAGE, Partition is the clear winner regardless of the overhead
+            # we'll see in unit tests.
+            | 'Split models into parallelizable PCollections' >> beam.Partition(
+                lambda m, _, kinds: kinds.index(job_utils.get_model_kind(m)),
+                # NOTE: Partition requires a hard-coded number of slices; it
+                # cannot be used with dynamic numbers generated in a pipeline.
+                # KIND_BY_INDEX is a constant tuple so that requirement is
+                # satisfied in this case.
+                len(KIND_BY_INDEX), KIND_BY_INDEX)
+        )
+
+        audit_error_pcolls = [
+            deleted_models
+            | 'Apply ValidateDeletedModel on deleted models' >> (
+                beam.ParDo(base_model_audits.ValidateDeletedModel()))
+        ]
+
+        model_groups = python_utils.ZIP(KIND_BY_INDEX, models_of_kind_by_index)
+        for kind, models_of_kind in model_groups:
+            # NOTE: Using extend() instead of append() because ApplyAuditDoFns
+            # produces an iterable of PCollections rather than a single one.
+            # NOTE: Label is missing because ApplyAuditDoFns labels itself.
+            audit_error_pcolls.extend(models_of_kind | ApplyAuditDoFns(kind))
+
+        return audit_error_pcolls | 'Combine audit results' >> beam.Flatten()
+
+
+class ApplyAuditDoFns(beam.PTransform):
+    """Runs every Audit DoFn targeting the models of a specific kind."""
+
+    def __init__(self, kind):
+        """Initializes a new ApplyAuditDoFns instance.
+
+        Args:
+            kind: str. The kind of models this PTransform will receive.
+        """
+        super(ApplyAuditDoFns, self).__init__(
+            label='Apply every Audit DoFn targeting %s' % kind)
+        self._kind = kind
+        self._do_fn_types = tuple(AUDIT_DO_FN_TYPES_BY_KIND[kind])
+
+    def expand(self, models_of_kind):
```

Comment:
This function name doesn't seem intuitive for the docstring (I would have expected something like get_audit_errors() or get_job_output()). In what sense is it an "expansion" -- is that describing the internal implementation more than the interface to the caller?

Reply:
This is the method that needs to be overridden by PTransform objects, so there's no way around it (unless we use a second function which I find less helpful).

Reply:
Ah I see, ok. Maybe just explain the connection to "expand" then.